src/serde/de.rs

```diff
@@ -0,0 +1,955 @@
+use crate::{IonError, IonResult, IonType, ReaderBuilder, StreamItem, StreamReader, Symbol};
+use chrono::{DateTime, FixedOffset};
+use serde::de;
+use serde::de::{DeserializeSeed, EnumAccess, MapAccess, SeqAccess, Visitor};
+use serde::Deserialize;
+use std::borrow::Cow;
+
+pub fn from_slice<'a, T>(s: &'a [u8]) -> IonResult<T>
+where
+    T: Deserialize<'a>,
+{
+    let mut deserializer = IonDeserializer {
+        reader: ReaderBuilder::new().build(s)?,
+    };
+    // Need to move twice
+    deserializer.reader.next()?; // Nothing header
+
+    let t = T::deserialize(&mut deserializer)?;
+
+    if deserializer.reader.ion_type().is_none() {
+        Ok(t)
+    } else {
+        Err(IonError::DecodingError {
+            description: "expected eof".to_string(),
+        })
+    }
+}
+
+pub fn from_str<'a, T>(s: &'a str) -> IonResult<T>
+where
+    T: Deserialize<'a>,
+{
+    let mut deserializer = IonDeserializer {
+        reader: ReaderBuilder::new().build(s)?,
+    };
+    deserializer.reader.next()?; // Nothing header
+
+    let t = T::deserialize(&mut deserializer)?;
+
+    if deserializer.reader.ion_type().is_none() {
+        Ok(t)
+    } else {
+        Err(IonError::DecodingError {
+            description: "expected eof".to_string(),
+        })
+    }
+}
+
+pub struct IonDeserializer<R> {
```

Comment:
Thinking this should just be named `Deserializer`. This is consistent to naming patterns in the serde ecosystem, e.g. [`serde_json::de::Deserializer`](https://docs.rs/serde_json/latest/serde_json/de/struct.Deserializer.html).

Reply:
Ack renaming

Reply:
Done rename